### ISSUE TITLE
use GridSearchCV  ,train loss change into nan

### ISSUE THREAD

**Issue by yysirs** [skorch-dev/skorch#743]
train code:
`net = NeuralNetClassifier(model,
                              # Shuffle training data on each epoch
                              iterator_train__shuffle=True)
    # 参数设置
    batch_size = [10]
    epochs = [30]
    learn_rate = [0.0001, 0.001]
    # momentum = [0.0, 0.2, 0.4, 0.6, 0.8, 0.9]
    # neurons = [5, 10, 20]
    # net.set_params(train_split=False, verbose=0)
    param_grid = dict(batch_size=batch_size, max_epochs=epochs, lr=learn_rate)
    # params = {
    #     'lr': [0.01, 0.02],
    #     'max_epochs': [10, 20],
    #     'module_num_units': [10, 20],
    # }
    # 网络搜索
    gs = GridSearchCV(net, param_grid, verbose=5, cv=5)
    gs.fit(X_train, y_train)  # 训练模型，训练集为采样后的数据`
model code 
`class BiLSTM_Model(nn.Module):
    def __init__(self, vocab_size, label_size, static_embeddings):
        super(BiLSTM_Model, self).__init__()
        self.hidden_dim = config.hidden_size
        self.embedding_dim = config.embedding_dim
        self.vocab_size = vocab_size
        self.label_size = label_size
        self.embedding = nn.Embedding(self.vocab_size, self.embedding_dim)
        # 这里使用static_embeddings来初始化embedding
        self.embedding.weight.data.copy_(torch.from_numpy(static_embeddings))
        self.BiLstm = nn.LSTM(self.embedding_dim, self.hidden_dim, num_layers=1, bidirectional=True, batch_first=True)
        self.dropout = nn.Dropout(config.dropout)
        self.linear1 = nn.Linear(self.hidden_dim*2, self.hidden_dim)
        self.linear2 = nn.Linear(self.hidden_dim, self.label_size)

    def forward(self, input):
        input_embedding = self.embedding(input)
        output, (h_n, c_n) = self.BiLstm(input_embedding)
        # 拼接正向最后一个输出和反向最后一个输出
        inputs = [h_n[-1, :, :], h_n[-2, :, :]]
        out = torch.cat(inputs, dim=-1)
        # 进行全连接
        out_fc1 = self.linear1(out)
        # 进行relu
        out_fc1_relu = F.relu(out_fc1)
        # 全连接
        out = self.linear2(out_fc1_relu)
        return F.log_softmax(out, dim=-1)`

when i train model without GridSearchCV, loss is not nan。

**Comment by yysirs** [comment]
<img width="436" alt="微信截图_20210304163105" src="https://user-images.githubusercontent.com/43271630/109945822-99e90280-7d12-11eb-9911-c0bfa1b0c8b9.png">
run images

**Comment by yysirs** [comment]
train code
`
net = NeuralNetClassifier(model,
                              # Shuffle training data on each epoch
                              iterator_train__shuffle=True)
    # 参数设置
    batch_size = [10]
    epochs = [30]
    learn_rate = [0.0001, 0.001]
    # momentum = [0.0, 0.2, 0.4, 0.6, 0.8, 0.9]
    # neurons = [5, 10, 20]
    # net.set_params(train_split=False, verbose=0)
    param_grid = dict(batch_size=batch_size, max_epochs=epochs, lr=learn_rate)
    # params = {
    #     'lr': [0.01, 0.02],
    #     'max_epochs': [10, 20],
    #     'module_num_units': [10, 20],
    # }
    # 网络搜索
    gs = GridSearchCV(net, param_grid, verbose=5, cv=5)
    gs.fit(X_train, y_train)  # 训练模型，训练集为采样后的数据
`

**Comment by yysirs** [comment]
`class BiLSTM_Model(nn.Module):
    def __init__(self, vocab_size, label_size, static_embeddings):
        super(BiLSTM_Model, self).__init__()
        self.hidden_dim = config.hidden_size
        self.embedding_dim = config.embedding_dim
        self.vocab_size = vocab_size
        self.label_size = label_size
        self.embedding = nn.Embedding(self.vocab_size, self.embedding_dim)
        # 这里使用static_embeddings来初始化embedding
        self.embedding.weight.data.copy_(torch.from_numpy(static_embeddings))
        self.BiLstm = nn.LSTM(self.embedding_dim, self.hidden_dim, num_layers=1, bidirectional=True, batch_first=True)
        self.dropout = nn.Dropout(config.dropout)
        self.linear1 = nn.Linear(self.hidden_dim*2, self.hidden_dim)
        self.linear2 = nn.Linear(self.hidden_dim, self.label_size)

    def forward(self, input):
        input_embedding = self.embedding(input)
        output, (h_n, c_n) = self.BiLstm(input_embedding)
        # 拼接正向最后一个输出和反向最后一个输出
        inputs = [h_n[-1, :, :], h_n[-2, :, :]]
        out = torch.cat(inputs, dim=-1)
        # 进行全连接
        out_fc1 = self.linear1(out)
        # 进行relu
        out_fc1_relu = F.relu(out_fc1)
        # 全连接
        out = self.linear2(out_fc1_relu)
        return F.log_softmax(out, dim=-1)`
model code

**Comment by BenjaminBossan** [comment]
The problem is probably that you're using `log_softmax`. `NeuralNetClassifier` expects `softmax` and applies the `log` for you (the reason being so that the outputs of the module are probabilities for the `predict_proba` method). Your options:

1. Change the final nonlinearity of your module to `softmax`
2. Don't use any nonlinearity and change the `criterion` of the net to `CrossEntropyLoss` (better numerical stability than 1.)

**Comment by yysirs** [comment]
i choose the first option, it's work. thank you!

**Comment by BenjaminBossan** [comment]
Glad to hear it.
